block/produce.go

```diff
@@ -148,14 +159,24 @@ func (m *Manager) produceBlock(allowEmpty bool, nextProposerHash *[32]byte) (*ty
 		return nil, nil, fmt.Errorf("load block: height: %d: %w: %w", newHeight, err, ErrNonRecoverable)
 	}
 
-	maxBlockDataSize := uint64(float64(m.Conf.BatchSubmitBytes) * types.MaxBlockSizeAdjustment)
-	proposerHashForBlock := [32]byte(m.State.Sequencers.ProposerHash())
-	// if nextProposerHash is set, we create a last block
-	if nextProposerHash != nil {
+	var (
+		maxBlockDataSize     = uint64(float64(m.Conf.BatchSubmitBytes) * types.MaxBlockSizeAdjustment)
+		proposerHashForBlock = [32]byte(m.State.Sequencers.ProposerHash())
+		nextProposerAddr     = m.State.Sequencers.Proposer.SettlementAddress
+		lastProposerBlock    = false // Indicates that the block is the last for the current seq. True during the rotation.
+	)
+	// if nextProposerInfo is set, we create a last block
+	if nextProposerInfo != nil {
 		maxBlockDataSize = 0
-		proposerHashForBlock = *nextProposerHash
+		proposerHashForBlock = nextProposerInfo.nextProposerHash
+		nextProposerAddr = nextProposerInfo.nextProposerAddr
+		lastProposerBlock = true
 	}
-	block = m.Executor.CreateBlock(newHeight, lastCommit, lastHeaderHash, proposerHashForBlock, m.State, maxBlockDataSize)
+	consensusMsgs, err := m.consensusMsgsOnCreateBlock(nextProposerAddr, lastProposerBlock)
+	if err != nil {
+		return nil, nil, fmt.Errorf("create consensus msgs for create block: last proposer block: %v, height: %d, next proposer addr: %s: %w: %w", lastProposerBlock, newHeight, nextProposerAddr, err, ErrNonRecoverable)
+	}
+	block = m.Executor.CreateBlock(newHeight, lastCommit, lastHeaderHash, proposerHashForBlock, m.State, maxBlockDataSize, consensusMsgs...)
```

Comment:
Ideally there should be only one point of adding consensus messages. Given they come from the ConsensusMessagesStream this should send them there.

The only problem I see here is that there is no implementation of the stream as of now. I would put a message somewhere with a TODO to unify it.